server/src/main/java/com/thoughtworks/go/server/newsecurity/x509/CachingSubjectDnX509PrincipalExtractor.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2018 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.newsecurity.x509;
+
+import net.sf.ehcache.*;
+import net.sf.ehcache.config.CacheConfiguration;
+import net.sf.ehcache.store.MemoryStoreEvictionPolicy;
+import org.springframework.dao.DataRetrievalFailureException;
+import org.springframework.security.ui.preauth.x509.SubjectDnX509PrincipalExtractor;
+import org.springframework.security.ui.preauth.x509.X509PrincipalExtractor;
+
+import java.security.cert.X509Certificate;
+
+public class CachingSubjectDnX509PrincipalExtractor implements X509PrincipalExtractor {
```

Comment:
Since parsing the X509 certificate to extract the subjectDN is a potentially expensive operation, we cache this parsing... Effectively a replacement for these lines: https://github.com/gocd/gocd/blob/7672cd956ef5c62f9957592ffc543d11509a31d6/server/webapp/WEB-INF/applicationContext-acegi-security.xml#L175-L178